### PR TITLE
IMU quaternion v4 

### PR DESCRIPTION
**Pull request by adrianmiriuta**
phase out rmat (use quaternion)

implemented static gyro noise integration lock  trough standard deviation modulus
improved gyro Integration (PCDM Acta Mech 224, 3091–3109 (2013))
separated from Mahony PI
corrected acc
corrected gyro
increased default dcm_ki dcm_kp for faster convergence
removed ATTITUDE task rescheduling

### CODE REVIEW

**Comment by brucesdad13** [comment]
Sweet 😁 Have flown this a bit on your betaflight fork with my whoop (an older version) and loved it.

**Comment by adrianmiriuta** [comment]
@brucesdad13 
thx !

it still  needs beta testing.

**Comment by kidBrazil** [comment]
@adrianmiriuta Thanks for your contribution. We are in the process of reviewing it still.

Could you tell me a little more about how much testing has gone into this so far?

**Comment by adrianmiriuta** [comment]
@kidBrazil 

it is indoor tested (working) on:
LUX_RACE 
REVO  
LUXV2_RACE
FF_FORTINIF4
OMNIBUSF4SD

Because of the manifold improvements in
Attitude, Gyro, Acc, Mag handling, ImuAHRS changes
it needs serious beta testing ! (from different testers)

It was submitted here about 2 months ago:
https://github.com/betaflight/betaflight/pull/5182
Then after long discussions it was put on hold.
Then it was eventually taken over here:
https://github.com/iNavFlight/inav/pull/2894
the functions where name changed, any similarities rejected ... (and so on)
Now the devs from BF argue to drop my PR and port the copy.
https://github.com/betaflight/betaflight/pull/5182#issuecomment-374572410

**Comment by kidBrazil** [comment]
@adrianmiriuta Ok, I will enlist some beta testers to do some thorough testing of your PR

**Comment by robert-b** [comment]
adrian,
have you checked the pi controller inside mahony?
how did you test this?
no overshoot?

**Comment by adrianmiriuta** [comment]
@robert-b 

PI controller response with minimal overshoot.

set debug_mode = IMU

debug[0]  = gyro absolute rotation speed
debug[1]  = PI controller  output
debug[2] = acc absoulute value
debug[3] = gyro static noise standard derivation (1 sigma)

tested with a non mounted LUX_RACE 
convergence time after gyro overflow (> 2000°/s)

**Comment by kidBrazil** [comment]
@adrianmiriuta - We have been conducting some testing all day. Results are promising so far, no issues reported - some users reported better feel in the air but hard to say if its something quantifiable.

CPU usage seems largely unchanged for most users as well.

Will report back when testing is fully concluded.

**Comment by robert-b** [comment]
yes, that was my thinking.
the correction with mahony is 2nd order in comparison with the madgwick linear-gradient variant.
i would reject the pr only because the p value is WAY to high.
you have to reduce it!

for example flying through a turbulent zone and the controller goes crazy.
else it looks nice.

**Comment by adrianmiriuta** [comment]
@kidBrazil 
cpu load is about 
2% higher on F3
1% higher on F4, F7
in my tests

**Comment by adrianmiriuta** [comment]
@robert-b 
I didn't noticed any oscillations in the PI output  (in the logs),
but I only tested indoor (only single flips and rolls ).
If this shows as a problem in the testing ,
I can reduce the default dcm  factors.

**Comment by robert-b** [comment]
adrian,
even light overshooting is bad.

**Comment by kidBrazil** [comment]
@robert-b @adrianmiriuta I had multiple users test this branch specifically today - They all tested with outdoor flying and I told them to really push the limits as much as they felt comfortable.

All have reported positive results, no one complained of issues with tune or feel. Most of them used Diff from previous build too so no retuning.

Our testing is concluded on this branch unless there are changes to the codebase.

File: src/main/common/maths.c [hunk 1]
```diff
@@ -345,3 +345,117 @@ int16_t qMultiply(fix12_t q, int16_t input) {
...
+    const float y = l->w * r->y - l->x * r->z + l->y * r->w + l->z * r->x;
+    const float z = l->w * r->z + l->x * r->y - l->y * r->x + l->z * r->w;
+    o->w = w;
+    o->x = x;
+    o->y = y;
+    o->z = z;
+}
+
+void quaternionNormalize(quaternion *q) {
+    float modulus = sqrtf(q->w * q->w + q->x * q->x + q->y * q->y + q->z * q->z);
```
Comment (LexioTech) [inline src/main/common/maths.c]: Could be just `float modulus = quaternionModulus(q);`

File: src/main/common/maths.c [hunk 2]
```diff
@@ -345,3 +345,117 @@ int16_t qMultiply(fix12_t q, int16_t input) {
...
+
+void quaternionCopy(quaternion *s, quaternion *d) {
+    d->w = s->w;
+    d->x = s->x;
+    d->y = s->y;
+    d->z = s->z;
+}
+
+void quaternionInverse(quaternion *i, quaternion *o) {
+    float norm = i->w * i->w + i->x * i->x + i->y * i->y + i->z * i->z;
```
Comment (LexioTech) [inline src/main/common/maths.c]: Could be just `float norm = quaternionNorm(i)`

File: src/main/common/maths.c [hunk 3]
```diff
@@ -345,3 +345,117 @@ int16_t qMultiply(fix12_t q, int16_t input) {
...
+float quaternionDotProduct(quaternion *l, quaternion *r) {
+    return(l->w * r->w + l->x * r->x + l->y * r->y + l->z * r->z);
+}
+
+float quaternionNorm(quaternion *q) {
+    return(q->w * q->w + q->x * q->x + q->y * q->y + q->z * q->z);
+}
+
+float quaternionModulus(quaternion *q) {
+    return(sqrtf(q->w * q->w + q->x * q->x + q->y * q->y + q->z * q->z));
```
Comment (LexioTech) [inline src/main/common/maths.c]: return(sqrtf(quaternionNorm(q)));

File: src/main/common/maths.c [hunk 4]
```diff
@@ -345,3 +345,117 @@ int16_t qMultiply(fix12_t q, int16_t input) {
...
+    qPout->wz = qIn->w * qIn->z;
+    qPout->xx = qIn->x * qIn->x;
+    qPout->xy = qIn->x * qIn->y;
+    qPout->xz = qIn->x * qIn->z;
+    qPout->yy = qIn->y * qIn->y;
+    qPout->yz = qIn->y * qIn->z;
+    qPout->zz = qIn->z * qIn->z;
+}
+
+void quaternionMultiply(quaternion *l, quaternion *r, quaternion *o) {
```
Comment (LexioTech) [inline src/main/common/maths.c]: We should skip the const floats and go for direct set:

`o->w = l->w * r->w - l->x * r->x - l->y * r->y - l->z * r->z;`
`o->x = l->w * r->x + l->x * r->w + l->y * r->z - l->z * r->y;`
`o->y = l->w * r->y - l->x * r->z + l->y * r->w + l->z * r->x;`
`o->z = l->w * r->z + l->x * r->y - l->y * r->x + l->z * r->w;`

Same thing, just more clear code.

**Comment by adrianmiriuta** [comment]
@LexioTech 
thx.

**Comment by robert-b** [comment]
doesn't make much sense.
macro optimization - which is done by the compiler ...

**Comment by adrianmiriuta** [comment]
@robert-b 
it's cleanup
it's good

**Comment by LexioTech** [comment]
@adrianmiriuta Thank you! 👍 
@robert-b Anything that cleans up the code and removes duplications makes sense. And if for some reason a math equation that is used in 3 places (and not via in a function) needs to be changed, then you need to change it in 3 places... not just 1 like this. 

Yes, the compiler is smart and will optimize this, but we also need to optimize the code for reading and coding friendliness.

**Comment by robert-b** [comment]
`void quaternionNormalize(quaternion *q) {
    float modulus = sqrtf(q->w * q->w + q->x * q->x + q->y * q->y + q->z * q->z);
    if (modulus == 0) {
      modulus = 0.0000001;
    }
    q->w /= modulus;
    q->x /= modulus;
    q->y /= modulus;
    q->z /= modulus;
}
`
you assign a double to a float ...
multiplications are faster than divisions - still remains true these days.

and above there is a medianfilter template.
on every line the WTF counter is incremented multiple times.

**Comment by adrianmiriuta** [comment]
@robert-b 

>  you assign a double to a float ...

you mean  ?
modulus = 0.0000001f;


> multiplications are faster than divisions - still remains true these days.
and above there is a medianfilter template.
on every line the WTF counter is incremented multiple times.

sorry I don't understand what you mean.

**Comment by robert-b** [comment]
@LexioTech 
above there is a medianfilter template.
on every line the WTF counter is incremented multiple times.
if you are consequent - then you need it to change.
not just mentioning minor issues ...

@adrianmiriuta 
the 'new' assigment is correct.
and rewrite the division using multiplications.

but the re-viewers and you did not recognize that the quaternions aft the assigment 'explodes ...
in case of the assigment the quaternion has to be re-initialized:

	// If quaternion has become inappropriately short or is nan reinit.
	// THIS SHOULD NEVER ACTUALLY HAPPEN
	if ((fabs(norm) < 1e-3) || (norm != norm))
	{
         o->w = 1.0f;
         o->x = 0;
         o->y = 0;
         o->z = 0;
	}
	else
		...

my 2 cents on a code review.

**Comment by adrianmiriuta** [comment]
@robert-b 

> and rewrite the division using multiplications.

How would you do that ?

>but the re-viewers and you did not recognize that the quaternions aft the assigment 'explodes ...
in case of the assigment the quaternion has to be re-initialized:

Now I understand.

this is only a zero division avoidance . (not my best implementation ...)

this function is also used to normalize vectors In a quaternion typeset storage.
` quaternionNormalize(vAcc); `
so your approach cannot be used

**Comment by robert-b** [comment]
hmm,
later on you need to work with the quaternion.
in your case the copter is def. dead.
// If quaternion has become inappropriately short or is nan reinit.
// THIS SHOULD NEVER ACTUALLY HAPPEN
if ((fabs(modulus) < 1e-3) || (modulus != modulus))
{
    o->w = 1.0f;
    o->x = 0;
    o->y = 0;
    o->z = 0;
}
else
{
    modulus = 1.0f / modulus;
    o->w *= modulus;
    o->x *= modulus;
    o->y *= modulus;
    o->z *= modulus;
}
and pls. use this snippet as it is.
the quaternion HAS to be reset when the length is invalid!

**Comment by adrianmiriuta** [comment]
No !

I use quaternionNormalize()  also on vectors in quaternion storage.

If a unit quaternion is that degraded,
that means that you have a massive problem with the Implementation ! (that is not the case)

If you want to monitor that you should write a extra function to do that !
you are welcome to do that, if you feel like it.

**Comment by robert-b** [comment]
:-)

**Comment by adrianmiriuta** [comment]
@kidBrazil 

Ok !
I will push some related cherry picks from BF also to the commit.

I used blackbox log pretty extensive to debug the code,
I didn't notice anything unusual.
Do you know what feature is broken ?

**Comment by kidBrazil** [comment]
@adrianmiriuta - I was sent this screenshot from the user who complained about BBE being broken. He says the LPF frequencies are not showing up properly. Could be entirely unrelated to this branch so I will investigate as well. 
w80

**Comment by adrianmiriuta** [comment]
@kidBrazil 
attention I introduced a bug with commit 2258d065175bc9f9282ccef3fe5bb854403ec500

@LexioTech 
undone change request 
https://github.com/ButterFlight/butterflight/pull/79#discussion_r177906818

this change is forbidden.
because you can use same quaternion as imput and output.
you need const w, x, y, z as buffer.

sorry I overlooked that.

**Comment by adrianmiriuta** [comment]
@kidBrazil 

I am not able to replicate the BBE issue.
Can you please ask the user for a diff. 

image

**Comment by adrianmiriuta** [comment]
@kidBrazil 

> I will get a diff for you so we can go over it. We will be making an RC version with quaternions soon so we can get even more testing done. 

Take your time, tough testing is a good thing.

> So far all feedback has been positive or "no change" which I believe is fine also.

That's not very comforting !

From the "diff" i take most of your testers fly FPV ?
All changes in the PR are related to Attitude handling
1. no loss gyro integration
2. statistical noise removal
3. better acc gyro calibration 
4. quick convergence after gyro over range.

Do you have testers who fly LOS ?
Bugs would affect HORIZON_MODE, ANGLE_MODE, HEADFREE_MODE, 
essentialy all modes that rely on Attitude quaternion .

**Comment by adrianmiriuta** [comment]
@kidBrazil
I see your testers fly with HELIOSPRING.
can you find out how high the static gyro drift "on the bench" is  °/minute (with 32k/32k) ?
and the value of debug[4] with set debug_mode=IMU ?

**Comment by kidBrazil** [comment]
@adrianmiriuta So, looks like after some investigation this issue with BBE only affects HELIOSPRING - Ornery has disabled the notch filtering ENTIRELY on that target which is why it doesn't show up.. sorry I missed that completely and sorry if I wasted any of your time.

That said I will look into the parser as well for BBE just to make sure.

Unfortunately I would say that 99% of my test pilots are doing FPV flights at the moment, I should be able to find some people willing to test LOS. I will mention specifically that testers should try it in one of the other Attitude modes as well.

Thanks again for the collab.

**Pull request closed without merging**